src/cpp/rtps/persistence/SQLite3PersistenceService.cpp

```diff
@@ -29,29 +29,138 @@ namespace eprosima {
+static int database_version(sqlite3* db)
+{
+    sqlite3_stmt* version_stmt;
+    int version = 0;
+    sqlite3_prepare_v2(db, "PRAGMA user_version;", -1, &version_stmt, NULL);
+    while (SQLITE_ROW == sqlite3_step(version_stmt))
+    {
+        version = sqlite3_column_int(version_stmt, 0);
+    }
+    sqlite3_finalize(version_stmt);
+    version_stmt = nullptr;
+    return version;
+}
+
+static int upgrade_version_1_to_version_2(sqlite3* db)
+{
+    const char* update_statement = R"(
+PRAGMA foreign_keys = ON;
+
+CREATE TABLE IF NOT EXISTS writers_states(
+    guid TEXT PRIMARY KEY, -- CHECK(guid REGEXP '([0-9a-fA-F]{1,2}\.){11}[0-9a-fA-F]{1,2}\|([0-9a-fA-F]{1,2}\.){3}[0-9a-fA-F]{1,2}'),
+    last_seq_num INTEGER CHECK(last_seq_num > 0)) WITHOUT ROWID;
+
+-- Populate the writers_states with information available in the old writers table. Note the assumption that
+-- last_seq_num would be the largest seq_num associated to a history cache is not always true
+INSERT INTO writers_states SELECT guid, MAX(seq_num) FROM writers GROUP BY guid;    
+
+CREATE TABLE IF NOT EXISTS writers_histories(
+    guid TEXT, -- CHECK(guid REGEXP '([0-9a-fA-F]{1,2}\.){11}[0-9a-fA-F]{1,2}\|([0-9a-fA-F]{1,2}\.){3}[0-9a-fA-F]{1,2}'),
+    seq_num INTEGER CHECK(seq_num > 0),
+    instance BLOB CHECK(length(instance)=16),
+    payload BLOB,
+    PRIMARY KEY(guid, seq_num DESC), 
+    FOREIGN KEY (guid)
+       REFERENCES writers_states(guid) ) WITHOUT ROWID;
+
+-- Copy the contents of writers to the new table
+INSERT INTO writers_histories SELECT guid, seq_num, instance, payload FROM writers;
+
+-- Remove the old table, since it is no longer required
+DROP TABLE writers;
+
+-- Once the upgrade has succeded add the version number
+PRAGMA user_version = 2;
+)";
+
+    return sqlite3_exec(db,update_statement,0,0,0);
+}
+
+static int upgrade(sqlite3* db, int from, int to)
+{
+    if (from == to)
+    {
+        return SQLITE_OK;
+    }
+
+    if ((from == 0 || from == 1) && to == 2)
+    {
+        return upgrade_version_1_to_version_2(db);
+    }
+
+    // unsupported upgrade path
+    logError(RTPS_PERSISTENCE, "Unsupported database upgrade from version " << from << " to version " << to);
+    return SQLITE_ERROR;
+}
+
+static sqlite3* open_or_create_database(const char* filename, bool update_schema)
 {
     sqlite3* db = NULL;
     int rc;
+    int version = 2;
 
     // Open database
-    int flags = SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE |
+    int flags = SQLITE_OPEN_READWRITE |
                 SQLITE_OPEN_FULLMUTEX | SQLITE_OPEN_SHAREDCACHE;
     rc = sqlite3_open_v2(filename, &db, flags, 0);
     if (rc != SQLITE_OK)
     {
-        sqlite3_close(db);
-        return NULL;
+        //probably the database does not exists. Create new and no need to upgrade schema
+        flags = SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE |
+                SQLITE_OPEN_FULLMUTEX | SQLITE_OPEN_SHAREDCACHE;
+        rc = sqlite3_open_v2(filename, &db, flags, 0);
```

Comment:
Why don't use the `SQLITE_OPEN_CREATE` flag directly it was as being done before?

Reply:
We want to avoid the upgrade on new databases. If the database does not exist and is created, it is created empty. there will be no schema version, so we will infer it is version '1', but the upgrade procedure will fail because there will be no tables on it.

I find this easier than trying a fail-free upgrade: if we try to open it without creating and it fails, but opening it with creation succeeds, it is a new one, will already be created with the appropriate schema, and we skip the upgrade (which is in the else clause)